transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollSocketChannel.java

```diff
@@ -111,6 +114,29 @@ protected AbstractEpollUnsafe newUnsafe() {
         return new EpollSocketChannelUnsafe();
     }
 
+    @Override
+    boolean doConnect0(SocketAddress remote) throws Exception {
+        if (Native.IS_SUPPORTING_TCP_FASTOPEN && config.isTcpFastOpenConnect()) {
+            ChannelOutboundBuffer outbound = unsafe().outboundBuffer();
+            outbound.addFlush();
+            Object curr;
+            if ((curr = outbound.current()) instanceof ByteBuf) {
+                ByteBuf initialData = (ByteBuf) curr;
+                long localFlushedAmount = doWriteOrSendBytes(
+                        initialData, (InetSocketAddress) remote, true);
+                if (localFlushedAmount > 0) {
+                    // We had a cookie and our fast-open proceeded. Remove written data
+                    // then continue with normal TCP operation.
+                    outbound.removeBytes(localFlushedAmount);
+                    return true;
+                }
+            }
+            // If no cookie is present, the write fails with EINPROGRESS and this call basically
```

Comment:
do we need to check that it really returned `EINPROGRESS` and if the failure is something else throw ?

Reply:
Yes, but not here, I think.